field/TinyLang/Field/UParser.hs

```diff
@@ -0,0 +1,258 @@
+analysis and grammar a bit.  To avoid any ambiguities between lexical
+analysis and grammar we will explicitly mark all lexical tokens.
+
+Please note that expressions are the designated start production for
+the parser.
+
+We are using the following convention
+
+* /tokens/ are enclosed in parentheses @""@.
+* /regular expressions/ are enclosed in square brackets @[]@.
+
+== Literals
+
+Literals are defined as follows
+
+@
+bool-literal ::= "T" | "F"
+vec-literal  ::= "{" (bool-literal ("," bool-literal)*)? "}"
+int-literal  ::= [0-9]+
+@
+
+== Keywords
+
+Here is a list of keywords used in the language.
+
+@
+keyword ::=
+  "and"
+  "assert"
+  "do"
+  "else"
+  "end"
+  "for"
+  "if"
+  "inv"
+  "let"
+  "neq0"
+  "not"
+  "or"
+  "then"
+  "to"
+  "unpack"
+  "xor"
+@
+
+== Identifiers
+
+To avoid a syntactic clash with /bool-literals/, identifiers for field
+variables start with a lower-case letter.  Identifiers for boolean
+variables are prefiex by @?@ and identifiers for vector variables are
+prefixed by @#@.  We will check if an identifier is not a keyword.
+
+@
+ident ::= ( "?" | "#" ) [a-z] ([a-z0-9_'])*
+@
+
+== Constants
+
+At the moment constants consist of literals only.
+
+@
+const ::=
+  bool-literal
+  int-literal
+  vec-literal
+@
+
+== Expressions
+
+As mentioned above @expr@ is the designated start production for the
+parser.
+
+@
+expr ::=
+  const
+  "(" expr ")"
+  ident
+  expr infix-op expr
+  prefix-op expr
+  expr "[" expr "]"
+  statement ";" expr
+  "if" expr "then" expr "else" expr
+
+infix-op ::=
+  "and"
+  "or"
+  "xor"
+  "=="
+  "<="
+  "<"
+  ">="
+  ">"
+  "+"
+  "-"
+  "*"
+  "/"
+
+prefix-op ::=
+  "not"
+  "neq0"
+  "neg"
+  "inv"
+  "unpack"
+@
+
+== Statement
+
+Statements are a bit odd at the moment, as they are neither
+expressions nor the usual statements.
+
+@
+statement ::=
+  "let" var "=" expr
+  "assert" expr
+  "for" var "=" int "to" int "do" statements "end"
+
+statements ::=
+  (statement (";" statement)*)?
+@
+
+== Operator Precedence
+  Precedence: "not" > "xor" > "and" > "or"  (but use parentheses anyway).
+  if-then-else has to be parenthesised unless it's at the very top.
+
+  Precedence for numeric operators is standard:  {neg,inv} > {*,/} > {+,- }.
+  Things like "neg inv 5" are illegal: use parentheses.
+
+  The code is based on the tutorial at
+  https://markkarpov.com/megaparsec/parsing-simple-imperative-language.html
+
+  See also https://markkarpov.com/megaparsec/megaparsec.html
+-}
+
+
+module TinyLang.Field.UParser where
+
+import           Control.Applicative.Combinators hiding (many, sepBy, some)
+import           Text.Megaparsec
+import           Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer      as L
+import           Data.Void
+import qualified Data.Set                        as S
+import           GHC.Unicode ( isLower
+                             , isDigit)
+
+type Parser = Parsec Void String
+
+type Identifier = String
+
+data Const
+  = CBool  Bool
```

Comment:
We use 4 spaces for indentation instead of 2 in Plutus, so I decided to keep that formatting rule in this project as well. Here's what I have in my `emacs` config:

```
(setq tab-stop-list (number-sequence 4 200 4))
(custom-set-variables
 '(haskell-indentation-layout-offset 4)
 '(haskell-indentation-left-offset 4)
 '(haskell-indentation-starter-offset 4)
 '(haskell-indentation-where-post-offset 4)
 '(haskell-indentation-where-pre-offset 4)
 '(tab-width 4)
 ...
)